### PR TITLE
Add yaml linting

### PR DESCRIPTION
**Pull request by viccuad**
## Description

This PR adds yaml linting to the dev/ scripts and gitlab, which lints for yaml files, excluding `deploy/helm/kubecf/*`, as those are helm templates that should be better linted with `helm lint`.

To achieve it, it adds a bazel target `//dev/linters:yamllint` to install the yamllint python module, and a python wrapper to call it. I haven't found any other way in Bazel to pull a python module and then call it.

It provides the script `dev/linters/yamllint.sh` that calls the python wrapper. And it updates `.gitlab-ci.yaml` to call this last script.

Bazel's pip_install() doesn't install pip, so this PR needs https://github.com/f0rmiga/bazel-docker-image/pull/3.

## How Has This Been Tested?

I have tested it locally by running `dev/linters/yamllint.sh`.

I have tested it against gitlab-ci by editing the `.gitlab-ci.yaml` to `zypper in pip`, and running `gitlab-runner exec docker lint`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code has security implications.
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by viccuad** [comment]
Linting the helm templates fail, eg:
```
./deploy/helm/kubecf/assets/operations/instance_groups/bits.yaml
  1:3       error    syntax error: expected the node content, but found '-'
```
We can't use `helm lint` as it only accepts `Chart.yml` as input. Another option would be to litter all templates with `# yamllint disable-line`, but that doesn't seem like a better option.

**Comment by f0rmiga** [comment]
https://github.com/SUSE/kubecf/pull/165 Should unblock this PR.

**Comment by viccuad** [comment]
Rebased on top of master to include https://github.com/SUSE/kubecf/pull/165.
Edited PR message to signify that this PR lints yaml files excluding `deploy/helm/kubecf/*`, as those are helm templates that should be better linted with `helm lint`

File: WORKSPACE [hunk 1]
```diff
@@ -13,6 +13,20 @@ external_binary(
     windows = project.shellcheck.platforms.windows,
 )
 
+http_archive(
+    name = "rules_python",
+    url = "https://github.com/bazelbuild/rules_python/releases/download/0.0.1/rules_python-0.0.1.tar.gz",
```
Comment (f0rmiga) [inline WORKSPACE]: We are controlling all the repository rules versions and hash sums on the top-level def.bzl. Check the `bazel_skylib` target:

https://github.com/SUSE/kubecf/blob/6e48b3fd04c40d2a64d5faeb33b2d6f73a194d32/WORKSPACE#L73-L77

https://github.com/SUSE/kubecf/blob/6e48b3fd04c40d2a64d5faeb33b2d6f73a194d32/def.bzl#L91-L94

File: WORKSPACE [hunk 2]
```diff
@@ -13,6 +13,20 @@ external_binary(
...
+    sha256 = "aa96a691d3a8177f3215b14b0edc9641787abaaa30363a080165d06ab65e1161",
+)
+load("@rules_python//python:pip.bzl", "pip_repositories","pip_import")
```
Comment (f0rmiga) [inline WORKSPACE]: Could you add a line break between each function call, please? Once we add `skylint`, it will ask the same thing. :)

File: dev/linters/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,14 @@
+load("@yamllint//:requirements.bzl", "all_requirements")
+load("@rules_python//python:defs.bzl", "py_binary")
+
+py_binary(
+    name = "yamllint",
+    srcs = ["//dev/linters:run_yamllint.py"],
+    data = [
```
Comment (f0rmiga) [inline dev/linters/BUILD.bazel]: Remove `data` completely, it's not being used.

File: dev/linters/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+        # "//dev/linters:setup.cfg",
+    ],
+    main = "run_yamllint.py",
+    python_version = "PY2",
+    deps = [
```
Comment (f0rmiga) [inline dev/linters/BUILD.bazel]: Unnecessary `[]`. Just do `deps = all_requirements`.

File: dev/linters/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    ] + all_requirements,
+)
```
Comment (f0rmiga) [inline dev/linters/BUILD.bazel]: Missing line break at the end of the file.

File: dev/linters/run_yamllint.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import os
+from yamllint import cli
+# import yamllint
```
Comment (f0rmiga) [inline dev/linters/run_yamllint.py]: Remove commented code.

File: doc/Contribute.md [hunk 7]
```diff
@@ -97,18 +97,20 @@ associated documentation, if we have any.
 
 ## Linting
 
-Currently only one linter is available:
+Currently 2 linters are available:
```
Comment (f0rmiga) [inline doc/Contribute.md]: [nit] Missing a comma after `Currently`.

File: doc/Contribute.md [hunk 8]
```diff
@@ -129,7 +131,7 @@ around the deployment manifest the ability to easily toggle various
 preset configurations, for example the use of eirini instead of diego
 as the application scheduler.
 
-### Features
+
```
Comment (f0rmiga) [inline doc/Contribute.md]: Why this change?

File: dev/linters/yamllint.sh [hunk 9]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+set -o errexit
+
+# shellcheck disable=SC2046
+# We want word splitting with find.
+bazel run //dev/linters:yamllint -- \
+      -d "{extends: relaxed, rules: {line-length: {max: 120}}}" \
+      --strict $(find . -not -path "./deploy/helm/kubecf/*" -name '*.yaml' -or -name '*.yml' )
```
Comment (f0rmiga) [inline dev/linters/yamllint.sh]: [nit] Missing type file only.
[Super nit] Extra space on the closing parenthesis.

Comment (f0rmiga) [reply dev/linters/yamllint.sh]: I think we still want to lint `values.yaml`.

Comment (f0rmiga) [reply dev/linters/yamllint.sh]: I'm wondering if it would worth running the `yamllint` on the rendered templates. It would require inverting the build step with the linting step on CI. Something like:

```
bazel run //dev/linters:yamllint -- ... <(helm template --name kubecf ...)
```

Ping @mook-as.

Comment (mook-as) [reply dev/linters/yamllint.sh]: Running linters on generated output can often be a bad idea (because we would often get generated output that would be different than how a human would write things from scratch, e.g. with very long lines).

File: dev/linters/BUILD.bazel [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+load("@rules_python//python:defs.bzl", "py_binary")
+
+py_binary(
+    name = "yamllint",
+    srcs = ["//dev/linters:run_yamllint.py"],
+    data = [
+        # "//dev/linters:setup.cfg",
+    ],
+    main = "run_yamllint.py",
+    python_version = "PY2",
```
Comment (mook-as) [inline dev/linters/BUILD.bazel]: We should probably be using Python3 throughout instead? I don't think we have any other Python dependencies at the moment, and Python 2 is going to EOL "soon". (And yamllint supports Python 3, as far as I can tell.)

File: dev/linters/yamllint.sh (hunk 9 above)
Comment (f0rmiga) [reply dev/linters/yamllint.sh]: In this case, we should aim to generate human-like manifests. A linter on the generated output would help us with achieving that. It's a different case than running a linter on, let's say, gRPC generated code, although I'd argue that it may be subject to some pre-compile checks. The reason why I believe we should aim to generate human-like manifests is that it makes inspection easy for everyone from devs to final customers.

File: dev/linters/BUILD.bazel (hunk 10 above)
Comment (f0rmiga) [reply dev/linters/BUILD.bazel]: +1

Both Python 2 and 3 are available on the Docker image (along with pip). I'll drop Python 2 as soon as it reaches EOL.

File: dev/linters/yamllint.sh (hunk 9 above)
Comment (viccuad) [reply dev/linters/yamllint.sh]: Good catch, I have added `deploy/helm/kubecf/values.*`
We could always lint before and after build.

**Comment by viccuad** [comment]
Many thanks for the review :). Also, I'm quite embarrassed by the simple mistakes I did.

I have fixed the typos and force-pushed with new changes, which I think it's gonna be cleaner:
- Using `git_repository` instead of `http_archive`, needed for python3 support, saving the commit sha in `def.bzl`
- pip_repositories and pip_import need to be evaluated at WORKSPACE-time, added commit comments. The rest can happily go on ./dev/linters/BUILD.bazel.

**Comment by viccuad** [comment]
Rebased on top of master to fix conflicts.

File: dev/linters/yamllint.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -o errexit
+
+# shellcheck disable=SC2046
+# We want word splitting with find.
+bazel run //dev/linters:yamllint -- \
```
Comment (mook-as) [inline dev/linters/yamllint.sh]: I'm having issues running this; I get:

```
$ bazel run //dev/linters:yamllint
Starting local Bazel server and connecting to it...
DEBUG: Rule 'rules_python' indicated that a canonical reproducible form can be obtained by modifying arguments shallow_since = "1573842889 -0500"
INFO: Analyzed target //dev/linters:yamllint (18 packages loaded, 210 targets configured).
INFO: Found 1 target...
INFO: Deleting stale sandbox base …/1902c1ccc22df327a938dca0d6a953d0/sandbox
Target //dev/linters:yamllint up-to-date:
  bazel-bin/dev/linters/yamllint
INFO: Elapsed time: 14.173s, Critical Path: 0.06s
INFO: 0 processes.
INFO: Build completed successfully, 4 total actions
INFO: Build completed successfully, 4 total actions
Traceback (most recent call last):
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/kubecf/dev/linters/run_yamllint.py", line 2, in <module>
    from yamllint import cli
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__yamllint_1_19_0/yamllint/cli.py", line 25, in <module>
    from yamllint import linter
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__yamllint_1_19_0/yamllint/linter.py", line 19, in <module>
    import yaml
  File "…/1902c1ccc22df327a938dca0d6a953d0/execroot/kubecf/bazel-out/k8-fastbuild/bin/dev/linters/yamllint.runfiles/yamllint_pypi__PyYAML_5_1_2/yaml/__init__.py", line 374
    class YAMLObject(metaclass=YAMLObjectMetaclass):
                              ^
SyntaxError: invalid syntax
```

This appears to be because:

```
$ cat bazel-bin/dev/linters/yamllint
#!/usr/bin/env python

…
```

This appears to be https://github.com/bazelbuild/bazel/issues/4815. We seem to need to change how bazel is run to fix this:

```suggestion
bazel run --incompatible_use_python_toolchains //dev/linters:yamllint -- \
```

Looking around, maybe my version of bazel (0.26) is too old?  Ah, yes, installing bazel 0.29 (the newest version in `obs://devel:tools:building`) fixed things.  All good, then.
